File: src/grove/update.py
"""Check for newer versions of grove on GitHub."""

from __future__ import annotations

import json
import threading
import time
from pathlib import Path

_CACHE_FILE = Path("~/.grove/update-check.json").expanduser()
_CHECK_INTERVAL = 86400  # 24 hours
_REPO = "nicksenap/grove"


def _parse_version(v: str) -> tuple[int, ...]:
    return tuple(int(x) for x in v.split("."))


def _fetch_latest() -> None:
    """Fetch latest version from GitHub and write to cache."""
    try:
        from urllib.request import Request, urlopen

        req = Request(
            f"https://api.github.com/repos/{_REPO}/releases/latest",
            headers={"Accept": "application/vnd.github.v3+json", "User-Agent": "grove"},
        )
        with urlopen(req, timeout=5) as resp:
            data = json.loads(resp.read())
            latest = data["tag_name"].lstrip("v")

        _CACHE_FILE.parent.mkdir(parents=True, exist_ok=True)
        _CACHE_FILE.write_text(json.dumps({"last_check": time.time(), "latest": latest}))
    except Exception:
        pass


def get_newer_version(current: str) -> str | None:
    """Return latest version if newer than *current*, else ``None``.

    Reads a local cache (~/.grove/update-check.json) and kicks off a background
    refresh when the cache is older than 24 h.  Never blocks and never raises.
    """
    try:
        latest: str | None = None
        last_check = 0.0

        if _CACHE_FILE.exists():
            cached = json.loads(_CACHE_FILE.read_text())
            latest = cached.get("latest")
            last_check = cached.get("last_check", 0.0)

        # Refresh in background when stale
        if time.time() - last_check >= _CHECK_INTERVAL:
            threading.Thread(target=_fetch_latest, daemon=True).start()

        if latest and _parse_version(latest) > _parse_version(current):
            return latest
    except Exception:
        pass

    return None

File: grove-go/internal/update/update.go
// Package update provides background version checking against GitHub releases.
package update

import (
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/nicksenap/grove/internal/config"
)

const (
	cacheTTL   = 24 * time.Hour
	releaseURL = "https://api.github.com/repos/nicksenap/grove/releases/latest"
	cacheFile  = "update-check.json"
)

type updateCache struct {
	Version   string `json:"version"`
	CheckedAt int64  `json:"checked_at"`
}

func cachePath() string {
	return filepath.Join(config.GroveDir(), cacheFile)
}

// GetNewerVersion returns the latest version if it's newer than current, or "".
// Never blocks, never raises. Kicks off a background refresh if cache is stale.
func GetNewerVersion(current string) string {
	cache := loadCache()

	// Kick off background refresh if stale
	if cache == nil || time.Since(time.Unix(cache.CheckedAt, 0)) > cacheTTL {
		go fetchAndCache()
	}

	if cache == nil || cache.Version == "" {
		return ""
	}

	if isNewer(cache.Version, current) {
		return cache.Version
	}
	return ""
}

func loadCache() *updateCache {
	data, err := os.ReadFile(cachePath())
	if err != nil {
		return nil
	}
	var cache updateCache
	if err := json.Unmarshal(data, &cache); err != nil {
		return nil
	}
	return &cache
}

func fetchAndCache() {
	version := fetchLatest()
	if version == "" {
		return
	}
	cache := updateCache{
		Version:   version,
		CheckedAt: time.Now().Unix(),
	}
	data, _ := json.Marshal(cache)
	os.WriteFile(cachePath(), data, 0o644)
}

func fetchLatest() string {
	client := &http.Client{Timeout: 5 * time.Second}
	resp, err := client.Get(releaseURL)
	if err != nil {
		return ""
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		return ""
	}

	var release struct {
		TagName string `json:"tag_name"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&release); err != nil {
		return ""
	}

	return strings.TrimPrefix(release.TagName, "v")
}

// isNewer returns true if latest > current using simple string comparison.
func isNewer(latest, current string) bool {
	latestParts := parseVersion(latest)
	currentParts := parseVersion(current)

	for i := 0; i < len(latestParts) && i < len(currentParts); i++ {
		if latestParts[i] > currentParts[i] {
			return true
		}
		if latestParts[i] < currentParts[i] {
			return false
		}
	}
	return len(latestParts) > len(currentParts)
}

func parseVersion(v string) []int {
	v = strings.TrimPrefix(v, "v")
	parts := strings.Split(v, ".")
	result := make([]int, len(parts))
	for i, p := range parts {
		fmt.Sscanf(p, "%d", &result[i])
	}
	return result
}
